tests/codex/utils/testsafeasynciter.nim

```diff
@@ -0,0 +1,389 @@
+              return failure("Error on item: " & $i)
+            elif i < 4:
+              return success($i)
+          return failure("Unexpected error"),
+        finishOnErr = false,
+      )
+      iter3 = await filter[string](
+        iter2,
+        proc(i: ?!string): Future[bool] {.async: (raises: [CancelledError]).} =
+          without i =? i, err:
+            if err.msg == "Error on item: 1":
+              return false
+            else:
+              return true
+          if i == "0":
+            return false
+          else:
+            return true,
+        finishOnErr = false,
+      )
+
+    var collectedSuccess: seq[string]
+    var collectedFailure: seq[string]
+
+    for fut in iter3:
+      without i =? (await fut), err:
+        collectedFailure.add(err.msg)
+        continue
+      collectedSuccess.add(i)
+
+    check:
+      collectedSuccess == @["3"]
+      collectedFailure == @["Error on item: 2", "Unexpected error"]
+      iter3.finished
+
+  test "Collecting errors on `filter` when finish on error is true":
+    let
+      iter1 = SafeAsyncIter[int].new(0 ..< 5)
+      iter2 = map[int, string](
+        iter1,
+        proc(i: ?!int): Future[?!string] {.async: (raises: [CancelledError]).} =
+          if i =? i:
+            if i == 3:
+              return failure("Error on item: " & $i)
+            elif i < 3:
+              return success($i)
+          return failure("Unexpected error"),
+        finishOnErr = false,
+      )
+      iter3 = await filter[string](
+        iter2,
+        proc(i: ?!string): Future[bool] {.async: (raises: [CancelledError]).} =
+          without i =? i, err:
+            if err.msg == "Unexpected error":
+              return false
+            else:
+              return true
+          if i == "0":
+            return false
+          else:
+            return true,
+      )
+
+    var collectedSuccess: seq[string]
+    var collectedFailure: seq[string]
+
+    for fut in iter3:
+      without i =? (await fut), err:
+        collectedFailure.add(err.msg)
+        continue
+      collectedSuccess.add(i)
+
+    check:
+      collectedSuccess == @["1", "2"]
+      # On error iterator finishes and returns the error of the item
+      # that caused the error = that's why we see it here
+      collectedFailure == @["Error on item: 3"]
+      iter3.finished
+
+  test "Should propagate cancellation error immediately":
+    let fut: Future[Option[?!string]].Raising([CancelledError]) =
+      Future[Option[?!string]].Raising([CancelledError]).init("testsafeasynciter")
+
+    let iter1 = SafeAsyncIter[int].new(0 ..< 5).delayBy(10.millis)
+    let iter2 = await mapFilter[int, string](
+      iter1,
+      proc(i: ?!int): Future[Option[?!string]] {.async: (raises: [CancelledError]).} =
+        if i =? i:
+          if (i < 3):
+            return some(success($i))
+        return await fut,
+    )
+
+    proc cancelFut(): Future[void] {.async.} =
+      await sleepAsync(100.millis)
```

Comment:
Hmm, not sure if we want to make this non-deterministic with a `sleep`, perhaps it's better to use another future for signaling, which makes things less prone to intermittent failures?

Reply:
Yes, perhaps sleep is a bit too risky in that test. There will be follow-up on removing remaining references to the old async iter, and then I get back to it and use something more deterministic. I will than merge this one.